hbase-client/src/main/java/org/apache/hadoop/hbase/client/ZKConnectionRegistry.java

```diff
@@ -133,13 +133,17 @@ private static void tryComplete(MutableInt remaining, HRegionLocation[] locs,
       ServerName.valueOf(snProto.getHostName(), snProto.getPort(), snProto.getStartCode()));
   }
 
-  private void getMetaRegionLocation(CompletableFuture<RegionLocations> future,
+  @VisibleForTesting
+  void getMetaRegionLocation(CompletableFuture<RegionLocations> future,
       List<String> metaReplicaZNodes) {
     if (metaReplicaZNodes.isEmpty()) {
       future.completeExceptionally(new IOException("No meta znode available"));
     }
     HRegionLocation[] locs = new HRegionLocation[metaReplicaZNodes.size()];
     MutableInt remaining = new MutableInt(locs.length);
+    // Do NOT use replicaid as index into locations array. The location set may not be complete
```

Comment:
I do not think use index instead of replicaId is the correct way to fix the problem here.
You should use replicaId, and the problem here is the size of the array is incorrect. You should find out the maximum replicaId and create the locs array with maxReplicaId + 1 as its size.

Reply:
Thanks for the review @Apache9 

I need help understanding here why returning an array with empty slots is the way to go instead? What you thinking? Is it that you are afraid that elsewhere we are using replicaid as index?

When you say find the max replicaid, are you suggesting the max replicaid present in the list metaReplicaZNodes (doable)?  Or are you thinking the array should be same as the configured replica count (hard -- hard to pass an active table schema reference down to ConnectionRegistry)?

Thanks for the fast review.

Reply:
Based on Duo and Stack's comments, I spend quite some time trying to understand how `locs` is being used. 
It is being used in 
https://github.com/apache/hbase/blob/master/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ZKConnectionRegistry.java#L162

It is used for initialize RegionLocations() object. 
Within RegionLocations(), all logic is there trying to allocate a location array with replica id as index.
https://github.com/apache/hbase/blob/master/hbase-client/src/main/java/org/apache/hadoop/hbase/RegionLocations.java#L74

So it seems that at least in this place, it is fine to do in this way. Otherwise, the logic is duplicated twice?

Reply:
Thanks @huaxiangsun . That helps (Its also used here https://github.com/apache/hbase/blob/master/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ZKConnectionRegistry.java#L187 but if error, we don't get to this point because of ArrayOutOfBoundsException).

Let me just do what @apache9 suggests. Looks like there is precedent around the handling of this situation. Let me follow the model (even if it means double work as you point out).

Will be back in a bit w/ new patch.

Reply:
The problem is for RegionLocations. As Huaxiang pointed out that in RegionLocations we will handle the out of order replicas, I think it is fine to do something like this here. Maybe we could use a List instead of an array? And add a comment here to say that "we do not care about the order of the replicas or if there are holes, the constructor of RegionLocations will handle this". And we could add a Constructor for RegionLoations to accept a List. And for the old constructor which accepts an array, just do this(Arrays.asList(locations));

Reply:
Let me try and do what you suggest later @apache9 with a list and pointers back to RegionLocations. Making an array that is maxReplicaId+1 is going to have the tryComplete do more work than was asked for (see 'remaining' param). Thanks for the help.